server/schedule/hot_net.go

```diff
@@ -0,0 +1,204 @@
+	hotRegionThreshold := uint64(float64(stores.TotalReadBytes()) / divisor)
+
+	if hotRegionThreshold < hotWriteRegionMinFlowRate {
+		hotRegionThreshold = hotWriteRegionMinFlowRate
+	}
+	return w.isNeedUpdateStatCache(region, hotRegionThreshold, ReadFlow)
+}
+
+func (w *HotSpotCache) isNeedUpdateStatCache(region *core.RegionInfo, hotRegionThreshold uint64, kind FlowKind) (bool, *core.RegionStat) {
+	var (
+		v       *core.RegionStat
+		value   interface{}
+		isExist bool
+	)
+	key := region.GetId()
+
+	switch kind {
+	case WriteFlow:
+		value, isExist = w.writeFlow.Peek(key)
+	case ReadFlow:
+		value, isExist = w.readFlow.Peek(key)
+	}
+	newItem := &core.RegionStat{
+		RegionID:       region.GetId(),
+		FlowBytes:      region.WrittenBytes,
+		LastUpdateTime: time.Now(),
+		StoreID:        region.Leader.GetStoreId(),
+		Version:        region.GetRegionEpoch().GetVersion(),
+		AntiCount:      hotRegionAntiCount,
+	}
+
+	if isExist {
+		v = value.(*core.RegionStat)
+		newItem.HotDegree = v.HotDegree + 1
+	}
+	switch kind {
+	case WriteFlow:
+		if region.WrittenBytes >= hotRegionThreshold {
+			return true, newItem
+		}
+	case ReadFlow:
+		if region.ReadBytes >= hotRegionThreshold {
+			return true, newItem
+		}
+	}
+	// smaller than hotReionThreshold
+	if !isExist {
+		return false, newItem
+	}
+	if v.AntiCount <= 0 {
+		return true, nil
+	}
+	// eliminate some noise
+	newItem.HotDegree = v.HotDegree - 1
+	newItem.AntiCount = v.AntiCount - 1
+	newItem.FlowBytes = v.FlowBytes
+	return true, newItem
+}
+
+// Update updates the cache.
+func (w *HotSpotCache) Update(key uint64, item *core.RegionStat, kind FlowKind) {
+	switch kind {
+	case WriteFlow:
+		if item == nil {
+			w.writeFlow.Remove(key)
+		} else {
+			w.writeFlow.Put(key, item)
+		}
+	case ReadFlow:
+		if item == nil {
+			w.readFlow.Remove(key)
+		} else {
+			w.readFlow.Put(key, item)
+		}
+	}
+}
+
+func (w *HotSpotCache) regionStats(kind FlowKind) []*core.RegionStat {
+	var elements []*cache.Item
+	switch kind {
+	case WriteFlow:
+		elements = w.writeFlow.Elems()
+	case ReadFlow:
+		elements = w.readFlow.Elems()
+	}
+	stats := make([]*core.RegionStat, len(elements))
+	for i := range elements {
+		stats[i] = elements[i].Value.(*core.RegionStat)
+	}
+	return stats
+}
+
+// RandomHotRegionFromStore random picks a hot region in specify store.
+func (w *HotSpotCache) RandomHotRegionFromStore(storeID uint64, kind FlowKind, hotThreshold int) *core.RegionStat {
+	stats := w.regionStats(kind)
+	for _, i := range rand.Perm(len(stats)) {
+		if stats[i].HotDegree > hotThreshold && stats[i].StoreID == storeID {
+			return stats[i]
+		}
+	}
+	return nil
+}
+
+func (w *HotSpotCache) isRegionHot(id uint64, hotThreshold int) bool {
+	if stat, ok := w.writeFlow.Peek(id); ok {
+		return stat.(*core.RegionStat).HotDegree >= hotThreshold
+	}
+	if stat, ok := w.readFlow.Peek(id); ok {
+		return stat.(*core.RegionStat).HotDegree >= hotThreshold
+	}
```

Comment:
What if the id is in both _writeFlow_ and _readFlow_, and in _writeFlow_ it's not hot but in _readFlow_ it's hot?

Reply:
It also returns true now, Do we need to distinguish them?

Reply:
I think it will return false.

Reply:
I think we should skip hot region in balance whatever it is hot read or write.

Reply:
got it.